File: modules/ar_method_evaluator.c
/**
 * @file ar_method_evaluator.c
 * @brief Implementation of the method evaluator module
 */

#include "ar_method_evaluator.h"
#include "ar_method.h"
#include "ar_method_ast.h"
#include "ar_instruction_ast.h"
#include "ar_expression_evaluator.h"
#include "ar_instruction_evaluator.h"
#include "ar_frame.h"
#include "ar_heap.h"
#include "ar_io.h"

/**
 * Internal structure for method evaluator
 */
struct ar_method_evaluator_s {
    const method_t *ref_method;                          /* Method to evaluate (borrowed reference) */
    ar_expression_evaluator_t *own_expr_evaluator;      /* Expression evaluator (owned) */
    instruction_evaluator_t *own_inst_evaluator;        /* Instruction evaluator (owned) */
};

/**
 * Creates a new method evaluator for a specific method
 */
ar_method_evaluator_t* ar_method_evaluator__create(
    const method_t *ref_method
) {
    if (!ref_method) {
        ar__io__error("ar_method_evaluator__create: method cannot be NULL");
        return NULL;
    }
    
    // Verify the method has a parsed AST
    const ar_method_ast_t *ast = ar__method__get_ast(ref_method);
    if (!ast) {
        ar__io__error("ar_method_evaluator__create: method has no parsed AST");
        return NULL;
    }
    
    // Allocate evaluator structure
    ar_method_evaluator_t *evaluator = AR__HEAP__MALLOC(sizeof(ar_method_evaluator_t), "method_evaluator");
    if (!evaluator) {
        ar__io__error("ar_method_evaluator__create: Failed to allocate evaluator");
        return NULL;
    }
    
    // Store reference to method
    evaluator->ref_method = ref_method;
    
    // Create stateless expression evaluator
    evaluator->own_expr_evaluator = ar__expression_evaluator__create_stateless();
    if (!evaluator->own_expr_evaluator) {
        AR__HEAP__FREE(evaluator);
        return NULL;
    }
    
    // Create stateless instruction evaluator
    evaluator->own_inst_evaluator = ar_instruction_evaluator__create_stateless(
        evaluator->own_expr_evaluator
    );
    if (!evaluator->own_inst_evaluator) {
        ar__expression_evaluator__destroy(evaluator->own_expr_evaluator);
        AR__HEAP__FREE(evaluator);
        return NULL;
    }
    
    return evaluator;
}

/**
 * Destroys a method evaluator
 */
void ar_method_evaluator__destroy(ar_method_evaluator_t *own_evaluator) {
    if (!own_evaluator) {
        return;
    }
    
    // Destroy owned evaluators
    if (own_evaluator->own_inst_evaluator) {
        ar_instruction_evaluator__destroy(own_evaluator->own_inst_evaluator);
    }
    if (own_evaluator->own_expr_evaluator) {
        ar__expression_evaluator__destroy(own_evaluator->own_expr_evaluator);
    }
    
    // Free the evaluator structure
    AR__HEAP__FREE(own_evaluator);
}

/**
 * Evaluates a method using the provided frame
 */
bool ar_method_evaluator__evaluate(
    ar_method_evaluator_t *mut_evaluator,
    const ar_frame_t *ref_frame
) {
    if (!mut_evaluator || !ref_frame) {
        ar__io__error("ar_method_evaluator__evaluate: evaluator and frame cannot be NULL");
        return false;
    }
    
    // Get the method's AST
    const ar_method_ast_t *method_ast = ar__method__get_ast(mut_evaluator->ref_method);
    if (!method_ast) {
        ar__io__error("ar_method_evaluator__evaluate: method has no parsed AST");
        return false;
    }
    
    // Get the instruction count
    size_t count = ar_method_ast__get_instruction_count(method_ast);
    if (count == 0) {
        // No instructions to execute
        return true;
    }
    
    // Execute each instruction in sequence
    for (size_t i = 0; i < count; i++) {
        // Get instruction by line number (1-based)
        const ar_instruction_ast_t *inst_ast = ar_method_ast__get_instruction(method_ast, i + 1);
        if (!inst_ast) {
            ar__io__error("ar_method_evaluator__evaluate: Failed to get instruction %zu", i + 1);
            return false;
        }
        
        // Evaluate the instruction with the frame
        bool result = ar_instruction_evaluator__evaluate_with_frame(
            mut_evaluator->own_inst_evaluator,
            inst_ast,
            ref_frame
        );
        
        if (!result) {
            ar__io__error("ar_method_evaluator__evaluate: Failed to evaluate instruction %zu", i + 1);
            return false;
        }
    }
    
    return true;
}

File: modules/ar_method_evaluator.zig
const std = @import("std");
const builtin = @import("builtin");
const ar_allocator = @import("ar_allocator.zig");
const c = @cImport({
    @cInclude("ar_method_evaluator.h");
    @cInclude("ar_log.h");
    @cInclude("ar_instruction_evaluator.h");
    @cInclude("ar_method_ast.h");
    @cInclude("ar_frame.h");
    @cInclude("ar_agency.h");
    @cInclude("stdio.h");
});

/// Internal structure for the method evaluator
const ar_method_evaluator_t = struct {
    ref_log: *c.ar_log_t,
    ref_agency: ?*c.ar_agency_t,
    own_instruction_evaluator: ?*c.ar_instruction_evaluator_t,
};

/// Creates a new method evaluator
export fn ar_method_evaluator__create(
    ref_log: ?*c.ar_log_t,
    ref_agency: ?*c.ar_agency_t
) ?*ar_method_evaluator_t {
    if (ref_log == null or ref_agency == null) return null;
    
    // Allocate evaluator
    const own_evaluator = ar_allocator.create(ar_method_evaluator_t, "method_evaluator") orelse return null;
    
    // Initialize fields
    own_evaluator.ref_log = ref_log.?;
    own_evaluator.ref_agency = ref_agency;
    
    // Create the instruction evaluator internally
    own_evaluator.own_instruction_evaluator = c.ar_instruction_evaluator__create(ref_log, ref_agency);
    if (own_evaluator.own_instruction_evaluator == null) {
        ar_allocator.free(own_evaluator);
        return null;
    }
    
    return own_evaluator; // Ownership transferred to caller
}

/// Destroys a method evaluator
export fn ar_method_evaluator__destroy(own_evaluator: ?*ar_method_evaluator_t) void {
    if (own_evaluator == null) {
        return;
    }
    
    // Destroy the owned instruction evaluator
    if (own_evaluator.?.own_instruction_evaluator != null) {
        c.ar_instruction_evaluator__destroy(own_evaluator.?.own_instruction_evaluator);
    }
    
    // Free the evaluator
    ar_allocator.free(own_evaluator);
}


/// Evaluates a method AST using the provided frame
export fn ar_method_evaluator__evaluate(
    ref_evaluator: ?*const ar_method_evaluator_t,
    ref_frame: ?*const c.ar_frame_t,
    ref_ast: ?*const c.ar_method_ast_t
) bool {
    // Validate inputs
    if (ref_evaluator == null or ref_frame == null or ref_ast == null) {
        return false;
    }
    
    // Get instruction count
    const instruction_count = c.ar_method_ast__get_instruction_count(ref_ast);
    
    // std.debug.print("DEBUG: Method has {} instructions\n", .{instruction_count});
    
    // If no instructions, succeed
    if (instruction_count == 0) {
        return true;
    }
    
    // Evaluate each instruction in sequence
    var line_no: usize = 1;
    while (line_no <= instruction_count) : (line_no += 1) {
        // Get the instruction
        const ref_instruction = c.ar_method_ast__get_instruction(ref_ast, line_no);
        if (ref_instruction == null) {
            // Format the error message with line number
            var buffer: [256]u8 = undefined;
            const formatted = std.fmt.bufPrintZ(&buffer, "Failed to retrieve instruction at line {d}", .{line_no}) catch {
                c.ar_log__error(ref_evaluator.?.ref_log, "Failed to retrieve instruction");
                return false;
            };
            c.ar_log__error(ref_evaluator.?.ref_log, formatted);
            return false;
        }
        
        // Evaluate the instruction with the frame
        const success = c.ar_instruction_evaluator__evaluate(ref_evaluator.?.own_instruction_evaluator, ref_frame, ref_instruction);
        if (!success) {
            // Log error with line number
            var buffer: [256]u8 = undefined;
            const formatted = std.fmt.bufPrintZ(&buffer, "Method evaluation failed at line {d}", .{line_no}) catch {
                c.ar_log__error(ref_evaluator.?.ref_log, "Method evaluation failed");
                return false;
            };
            c.ar_log__error(ref_evaluator.?.ref_log, formatted);
            return false;
        }
    }
    
    // All instructions evaluated successfully
    return true;
}